File: android/kit/src/main/kotlin/org/secretary/sync/UniffiVaultSyncPort.kt
package org.secretary.sync

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import uniffi.secretary.SyncOutcomeDto
import uniffi.secretary.SyncStatusDto
import uniffi.secretary.VaultException
import uniffi.secretary.VetoDecisionDto
import uniffi.secretary.syncCommitDecisions
import uniffi.secretary.syncStatus
import uniffi.secretary.syncVault

/**
 * The real [VaultSyncPort] over the generated `uniffi.secretary` sync calls. This is the only
 * [VaultSyncPort] implementation that invokes the bindings (the pure DTO/error mappers reference
 * the generated types too, but only as translators). A faithful Kotlin mirror of iOS
 * `UniffiVaultSyncPort.swift`.
 *
 * [sync] and [commitDecisions] re-open the vault from the password (full Argon2id), so they run
 * on [ioDispatcher] (default [Dispatchers.IO]) to keep the caller responsive; [status] is a cheap
 * disk read and runs inline. The password [ByteArray] is forwarded per call and never retained.
 *
 * The three FFI functions are injectable seams defaulting to the real bindings, so the adapter's
 * wiring is host-testable with fakes (no native library loaded). Production code uses all defaults.
 */
class UniffiVaultSyncPort(
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
    private val statusFn: (String, ByteArray) -> SyncStatusDto = ::syncStatus,
    private val syncFn: (String, String, ByteArray, ULong) -> SyncOutcomeDto = ::syncVault,
    // (stateDir, vaultFolder, password, decisions, manifestHash, nowMs) -> outcome
    private val commitFn: (String, String, ByteArray, List<VetoDecisionDto>, ByteArray, ULong) -> SyncOutcomeDto =
        ::syncCommitDecisions,
) : VaultSyncPort {

    override suspend fun status(stateDir: String, vaultUuid: ByteArray): SyncStatus =
        mapStatus(callMappingErrors { statusFn(stateDir, vaultUuid) })

    override suspend fun sync(
        stateDir: String,
        vaultFolder: String,
        password: ByteArray,
        nowMs: ULong,
    ): SyncOutcome = withContext(ioDispatcher) {
        mapOutcome(callMappingErrors { syncFn(stateDir, vaultFolder, password, nowMs) })
    }

    override suspend fun commitDecisions(
        stateDir: String,
        vaultFolder: String,
        password: ByteArray,
        decisions: List<SyncVetoDecision>,
        manifestHash: ByteArray,
        nowMs: ULong,
    ): SyncOutcome = withContext(ioDispatcher) {
        val dtoDecisions = decisions.map(::toVetoDecisionDto)
        mapOutcome(callMappingErrors {
            commitFn(stateDir, vaultFolder, password, dtoDecisions, manifestHash, nowMs)
        })
    }
}

/** Run an FFI call, translating any [VaultException] into the domain [VaultSyncError]. */
private inline fun <T> callMappingErrors(block: () -> T): T =
    try {
        block()
    } catch (e: VaultException) {
        throw mapVaultSyncError(e)
    }

private fun toVetoDecisionDto(d: SyncVetoDecision): VetoDecisionDto =
    VetoDecisionDto(recordUuidHex = d.recordUuidHex, keepLocal = d.keepLocal)

File: ios/SecretaryKit/Sources/SecretaryKit/VaultAccess/UniffiVaultSyncPort.swift
import Foundation
import SecretaryVaultAccess

/// Real `VaultSyncPort` over the uniffi sync functions (#187). `sync` and
/// `commitDecisions` re-open the identity from the password (full Argon2id), so
/// they run off the calling actor via `runOffMainActor` — exactly like
/// `UniffiVaultOpenPort` — keeping a `@MainActor` caller responsive. `status`
/// is a cheap disk read and runs inline (no `runOffMainActor`): under
/// `SyncCoordinator` (a plain `actor`) that still executes on the cooperative
/// pool, never the main thread.
public struct UniffiVaultSyncPort: VaultSyncPort {
    public init() {}

    public func status(stateDir: String, vaultUuid: [UInt8]) async throws -> SyncStatus {
        do {
            let dto = try SecretaryKit.syncStatus(stateDir: stateDir, vaultUuid: Data(vaultUuid))
            return Self.mapStatus(dto)
        } catch let e as VaultError {
            throw mapVaultSyncError(e)
        }
    }

    public func sync(stateDir: String, vaultFolder: String,
                     password: [UInt8], nowMs: UInt64) async throws -> SyncOutcome {
        try await runOffMainActor {
            do {
                let dto = try withZeroizingData(password) { pw in
                    try SecretaryKit.syncVault(
                        stateDir: stateDir, vaultFolder: vaultFolder,
                        password: pw, nowMs: nowMs)
                }
                return Self.mapOutcome(dto)
            } catch let e as VaultError {
                throw mapVaultSyncError(e)
            }
        }
    }

    public func commitDecisions(stateDir: String, vaultFolder: String,
                                password: [UInt8], decisions: [SyncVetoDecision],
                                manifestHash: [UInt8], nowMs: UInt64) async throws -> SyncOutcome {
        try await runOffMainActor {
            do {
                let dtoDecisions = decisions.map {
                    VetoDecisionDto(recordUuidHex: $0.recordUuidHex, keepLocal: $0.keepLocal)
                }
                let dto = try withZeroizingData(password) { pw in
                    try SecretaryKit.syncCommitDecisions(
                        stateDir: stateDir, vaultFolder: vaultFolder, password: pw,
                        decisions: dtoDecisions, manifestHash: Data(manifestHash), nowMs: nowMs)
                }
                return Self.mapOutcome(dto)
            } catch let e as VaultError {
                throw mapVaultSyncError(e)
            }
        }
    }

    // MARK: - DTO → pure value type mapping

    private static func mapStatus(_ s: SyncStatusDto) -> SyncStatus {
        SyncStatus(
            hasState: s.hasState,
            deviceClocks: s.deviceClocks.map {
                DeviceClock(deviceUuidHex: $0.deviceUuidHex, counter: $0.counter)
            },
            lastStateWriteMs: s.lastStateWriteMs)
    }

    private static func mapVeto(_ v: VetoDto) -> SyncVeto {
        SyncVeto(recordUuidHex: v.recordUuidHex, recordType: v.recordType, tags: v.tags,
                 fieldNames: v.fieldNames, localLastModMs: v.localLastModMs,
                 peerTombstonedAtMs: v.peerTombstonedAtMs, peerDeviceHex: v.peerDeviceHex)
    }

    private static func mapCollision(_ c: CollisionDto) -> SyncCollision {
        SyncCollision(recordUuidHex: c.recordUuidHex, fieldNames: c.fieldNames)
    }

    private static func mapOutcome(_ o: SyncOutcomeDto) -> SyncOutcome {
        switch o {
        case .nothingToDo:          return .nothingToDo
        case .appliedAutomatically: return .appliedAutomatically
        case .silentMerge:          return .silentMerge
        case .mergedClean:          return .mergedClean
        case .rollbackRejected:     return .rollbackRejected
        case let .conflictsPending(vetoes, collisions, manifestHash):
            return .conflictsPending(
                vetoes: vetoes.map(mapVeto),
                collisions: collisions.map(mapCollision),
                manifestHash: [UInt8](manifestHash))
        }
    }
}
